hashCheck/hashCheck.go

```diff
@@ -0,0 +1,230 @@
+package hashCheck
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/sonalys/fake"
+	"golang.org/x/tools/go/packages"
+	"io"
+	"os"
+	"path/filepath"
+)
+
+const (
+	fileName = "fake.lock.json"
+)
+
+func CompareFileHashes(inputDirs, ignore []string) ([]string, error) {
+	res := make([]string, 0)
+
+	for _, dir := range inputDirs {
+
+		files, err := fake.ListGoFiles(dir, ignore)
+		if err != nil {
+			return nil, fmt.Errorf("getFiles: %w", err)
+		}
+
+		groups := groupByDirectory(files)
+
+		for group, data := range groups {
+			model, err := parseJsonModel(group)
+
+			if err != nil {
+				return nil, fmt.Errorf("parseJsonModel: %w", err)
+			}
+
+			for _, file := range data {
+
+				imports, err := getPackagesGosum(file)
+				if err != nil {
+					return nil, fmt.Errorf("getImportPath: %w", err)
+				}
+
+				goSum, err := hashFiles(imports...)
+				if err != nil {
+					return nil, fmt.Errorf("hashFiles: %w", err)
+				}
+
+				hash, err := hashFiles(file)
+				if err != nil {
+					return nil, fmt.Errorf("hashFiles: %w", err)
+				}
+
+				if model[file].Hash != hash || model[file].GoSum != goSum {
+					res = append(res, file)
+				}
+
+				model[file] = FileHashData{
+					Hash:  hash,
+					GoSum: goSum,
+				}
+
+			}
+			err = saveHashToFile(dir, group, model)
+			if err != nil {
+				return nil, fmt.Errorf("saveHashToFile: %w", err)
+			}
+
+		}
+
+	}
+
+	return res, nil
+}
+
+/*
+groupByDirectory groups files by their directory
+Example:
+
+Input:
+
+	files := []string{
+		"/home/user/documents/file1.txt",
+		"/home/user/documents/file2.txt",
+		"/home/user/images/image1.png",
+		"/home/user/images/image2.png",
+		"/home/user/images/image3.png",
+	}
+
+Output:
+
+		{
+		"/home/user/documents": []string{
+			"/home/user/documents/file1.txt",
+			"/home/user/documents/file2.txt",
+		},
+		"/home/user/images": []string{
+			"/home/user/images/image1.png",
+			"/home/user/images/image2.png",
+			"/home/user/images/image3.png",
+		},
+	}
+*/
+func groupByDirectory(files []string) map[string][]string {
+	groups := make(map[string][]string)
+	for _, file := range files {
+		dir := filepath.Dir(file)
+		groups[dir] = append(groups[dir], file)
+	}
+	return groups
+}
+
+// parseJsonModel reads and parses the json model from the fake.lock.json file
+func parseJsonModel(path string) (Hashes, error) {
+	data, err := os.ReadFile(filepath.Join("mocks", path, fileName))
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return Hashes{}, nil
+		}
+		return nil, err
+	}
+
+	var model Hashes
+	err = json.Unmarshal(data, &model)
+	if err != nil {
+		return nil, err
+	}
+
+	return model, nil
+}
+
+/*
+getPackagesGosum takes path to a .go file
+returns a list of go.sum files for the given file's dependencies
+*/
+func getPackagesGosum(file string) ([]string, error) {
+	res := make([]string, 0)
```

Comment:
Can be moved to line 145 so you know the final size of the slice before initializing.